### PR TITLE
ddl(ticdc): add charset and collate to ddl event

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve?

Issue Number: close #8568 

### What is changed and how it works?

* add charset and collation to the DDL event.

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

1. deploy 2 tidb clusters, as upstream and downstream.
2. create a changefeed
3. run the following SQL

```
set character_set_connection=utf8;

CREATE TABLE `t` (
  `pk` bigint(20) NOT NULL AUTO_INCREMENT,
  `j` json DEFAULT NULL,
  `i` int(11) DEFAULT NULL,
  `c` char(64) DEFAULT NULL,
  PRIMARY KEY (`pk`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=30001;

alter table t add index j((cast(j->'$.number[*]' as signed array))); 

show create table t;
```

4. both tidb clusters have the same output.

### Release note 

```release-note
`None`
```

### CODE REVIEW

**Comment by 3AceShowHand** [comment]
/hold

This PR waiting for the https://github.com/pingcap/tidb/pull/42750 get merged

**Comment by 3AceShowHand** [comment]
image

This need further investigation, the table structure matched after we set the charset and collation into the DDL, but the table data is incorrect. 

```
Error in comparison process:
Error 1054 (42S22): Unknown column '_V$_j_0' in 'where clause' error occured in `test`.`t`
```

[sync_diff.log](https://github.com/pingcap/tiflow/files/11169989/sync_diff.log)

**Comment by 3AceShowHand** [comment]
Screenshot 2023-04-07 at 14 00 13

sync-diff passed

File: cdc/owner/ddl_sink.go [hunk 1]
```diff
@@ -352,13 +352,14 @@ func (s *ddlSinkImpl) isInitialized() bool {
 }
 
 // addSpecialComment translate tidb feature to comment
-func addSpecialComment(ddlQuery string) (string, error) {
-	stms, _, err := parser.New().ParseSQL(ddlQuery)
+func addSpecialComment(ddl *model.DDLEvent) (string, error) {
+	stms, _, err := parser.New().Parse(ddl.Query, ddl.Charset, ddl.Collate)
```
Comment (overvenus) [inline cdc/owner/ddl_sink.go:356]: What if `ddl.Charset` is nil?

File: cdc/owner/ddl_sink.go [hunk 2]
```diff
@@ -376,5 +377,13 @@ func addSpecialComment(ddlQuery string) (string, error) {
...
 		return "", errors.Trace(err)
 	}
-	return sb.String(), nil
+
+	result := sb.String()
+	log.Info("add special comment to DDL",
+		zap.String("DDL", ddl.Query),
+		zap.String("charset", ddl.Charset),
+		zap.String("collate", ddl.Collate),
+		zap.String("result", result))
```
Comment (overvenus) [inline cdc/owner/ddl_sink.go:386]: Please add changefeed and namespace.

File: cdc/owner/ddl_sink.go (hunk 1 above)
Comment (3AceShowHand) [reply cdc/owner/ddl_sink.go:356]: It may happen if the TiDB is in the old version compared to the TiCDC, so the charset is "", so it's ignored.

**Review by sdojjy — APPROVED** [review]

File: cdc/owner/ddl_sink.go (hunk 1 above)
Comment (3AceShowHand) [reply cdc/owner/ddl_sink.go:356]: It's the default charset and collate, `utf8mb4`

**Review by overvenus — APPROVED** [review]

**Pull request merged**